README.md
# Package for Laravel calculating Co2 emission based on Sustainable Web Design (SWD) model

[![Latest Version on Packagist](https://img.shields.io/packagist/v/ahoicloud/co2.svg?style=flat-square)](https://packagist.org/packages/ahoicloud/co2)
[![GitHub Tests Action Status](https://img.shields.io/github/actions/workflow/status/ahoicloud/co2/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/ahoicloud/co2/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/actions/workflow/status/ahoicloud/co2/fix-php-code-style-issues.yml?branch=main&label=code%20style&style=flat-square)](https://github.com/ahoicloud/co2/actions?query=workflow%3A"Fix+PHP+code+style+issues"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/ahoicloud/co2.svg?style=flat-square)](https://packagist.org/packages/ahoicloud/co2)

Every byte of data that’s uploaded or downloaded produces CO2. By being able to calculate these emissions, developers can enabled themself and their users to create more efficient, lower carbon decisions. Here are a few examples:
+ Create a carbon budget for your site
+ Inform the user when they uploading or downloading carbon intensive files

The package uses the [Sustainable Web Design (SWD)](https://sustainablewebdesign.org/calculating-digital-emissions) model.
The yearly average grid intensity data comes from [Ember](https://ember-climate.org/data/data-explorer/), as well as marginal intensity data from the [UNFCCC](https://unfccc.int/) (United Nations Framework Convention on Climate Change).


The package is build for [Laravel](https://github.com/laravel/framework) and is based on the great  work done by [The Green Web Foundation](https://thegreenwebfoundation.org) in the  [CO2.js](https://github.com/thegreenwebfoundation/co2.js) package.



## Installation

You can install the package via composer:

```bash
composer require ahoicloud/co2
```

### Optional 

You can publish the config file with:

```bash
php artisan vendor:publish --tag="co2-config"
```

This is the contents of the published config file:

```php
return [

    'KWH_PER_GB' => 0.81,
    'END_USER_DEVICE_ENERGY' => 0.52,
    'NETWORK_ENERGY' => 0.14,
    'DATACENTER_ENERGY' => 0.15,
    'PRODUCTION_ENERGY' => 0.19,
    'GLOBAL_GRID_INTENSITY' => 442,
    'RENEWABLES_GRID_INTENSITY' => 50,
    'FIRST_TIME_VIEWING_PERCENTAGE' => 0.75,
    'RETURNING_VISITOR_PERCENTAGE' => 0.25,
    'PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD' => 0.02,

];
```


## Usage
### Total energy use by each corresponding system component in kilowatt hours
```php
$co2 = new Ahoicloud\Co2();
echo $co2->energyPerByteByComponent($byte);
```
Or as a facade

```php
use Ahoicloud\Co2\Facades\Co2;
echo Co2::energyPerByteByComponent($byte);
```


### Total co2 figures key by the each component
```php
$co2 = new Ahoicloud\Co2();
echo $co2->co2byComponent($byte);
```
Or as a facade

```php
use Ahoicloud\Co2\Facades\Co2;
echo Co2::co2byComponent($byte);
```

### Total co2 figures key by the each component
```php
$co2 = new Ahoicloud\Co2();
echo $co2->perByte($byte);
```
Or as a facade

```php
use Ahoicloud\Co2\Facades\Co2;
echo Co2::perByte($byte);
```
## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Arne Breitsprecher](https://github.com/ahoicloud)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

The code for [thegreenwebfoundation/co2.js](https://github.com/thegreenwebfoundation/co2.js) is licensed Apache 2.0. ([What does this mean?](https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)))

The average carbon intensity data from Ember is published under the Creative Commons ShareAlike Attribution Licence ([CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)). ([What does this mean?](https://tldrlegal.com/license/creative-commons-attribution-sharealike-4.0-international-(cc-by-sa-4.0)))

The marginal intensity data is published by the Green Web Foundation, under the Creative Commons ShareAlike Attribution Licence ([CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)). ([What does this mean?](https://tldrlegal.com/license/creative-commons-attribution-sharealike-4.0-international-(cc-by-sa-4.0)))

See LICENCE for more.


CHANGELOG.md
# Changelog

All notable changes to `co2` will be documented in this file.

## v1.0.2 - 2023-02-04

Just adjusting the descriptions of composer.json and README.md

**Full Changelog**: https://github.com/ahoicloud/co2/compare/v1.0.1...v1.0.2

## v1.0.1 - 2023-02-04

##Updated
Updated the README.md file

**Full Changelog**: https://github.com/ahoicloud/co2/compare/v.1.0.0...v1.0.1

## First release  - 2023-02-04

This is the first package release for the co2 calculator

### What's Changed

- Bump aglipanci/laravel-pint-action from 1.0.0 to 2.1.0 by @dependabot in https://github.com/ahoicloud/co2/pull/1

**Full Changelog**: https://github.com/ahoicloud/co2/commits/v.1.0.0


tests/Pest.php
<?php

use Ahoicloud\Co2\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/Co2Test.php
<?php

use Ahoicloud\Co2\Co2;

const averageWebsiteInBytes = 2257715.2;
const MILLION = 1000000;
const MILLION_GREY = 0.35802;
const MILLION_GREEN = 0.31039;
const MILLION_GREY_DEVICES = 0.18617;
const MILLION_GREY_NETWORKS = 0.05012;
const MILLION_GREY_DATACENTERS = 0.0537;
const MILLION_GREEN_DATACENTERS = 0.006075;
const MILLION_GREY_PRODUCTION = 0.06802;

it('should return a array with numbers for each system component', function () {
    // Compare these with the carbon intensity tab C7-C10.
    // https://docs.google.com/spreadsheets/d/1eFlHhSBus_HqmoXqX237eAYr0PREUhTr6YBxznQC4jI/edit#gid=0
    $co2 = new Co2();
    // These numbers should match the spreadsheet
    $groupedEnergy = $co2->energyPerByteByComponent(averageWebsiteInBytes);
    expect($groupedEnergy['consumerDeviceEnergy'])->toBe(0.0009509496422400002);
    expect($groupedEnergy['networkEnergy'])->toBe(0.00025602490368000007);
    expect($groupedEnergy['dataCenterEnergy'])->toBe(0.0002743123968);
    expect($groupedEnergy['productionEnergy'])->toBe(0.00034746236928000007);
});

it('should return a number in kilowatt hours for the given data transfer in bytes', function () {
    $co2 = new Co2();
    $energyForTransfer = $co2->energyPerByte(averageWebsiteInBytes);
    expect($energyForTransfer)->toBe(0.0018287493120000002);
});
it('should return a single number for CO2 emissions', function () {
    $co2 = new Co2();
    expect($co2->perByte(2257715.2))->toBeNumeric();
});

it('should return a array for CO2 emissions', function () {
    $co2 = new Co2();

    expect($co2->energyPerByteByComponent(2257715.2))->toBeArray();
});
it('returns a CO2 number for data transfer', function () {
    $co2 = new Co2();
    expect(number_format($co2->perByte(MILLION), 5))->toBe(number_format(MILLION_GREY, 5));
});
it("returns a lower CO2 number for data transfer from domains using entirely 'green' power", function () {
    $co2 = new Co2();
    expect(number_format($co2->perByte(MILLION), 5))->toBe(number_format(MILLION_GREY, 5));
    expect(number_format($co2->perByte(MILLION, true), 5))->toBe(number_format(MILLION_GREEN, 5));
});

it('returns adjusted data center and total emissions for when green, other values remain the same as grey', function () {
    $co2 = new Co2();
    $co2_result_segements = $co2->perByte(MILLION, true, true);

    expect(number_format($co2_result_segements['consumerDeviceCO2'], 5))->toBe(number_format(MILLION_GREY_DEVICES, 5));
    expect(number_format($co2_result_segements['dataCenterCO2'], 6))->toBe(number_format(MILLION_GREEN_DATACENTERS, 6));
    expect(number_format($co2_result_segements['total'], 5))->toBe(number_format(MILLION_GREEN, 5));
});

it('returns an object with devices, networks, data centers, and production emissions shown separately, as well as the total emissions', function () {
    $co2 = new Co2();
    $co2_result_segements = $co2->perByte(MILLION, false, true);

    expect(number_format($co2_result_segements['consumerDeviceCO2'], 5))->toBe(number_format(MILLION_GREY_DEVICES, 5));
    expect(number_format($co2_result_segements['dataCenterCO2'], 5))->toBe(number_format(MILLION_GREY_DATACENTERS, 5));
    expect(number_format($co2_result_segements['networkCO2'], 5))->toBe(number_format(MILLION_GREY_NETWORKS, 5));
    expect(number_format($co2_result_segements['total'], 5))->toBe(number_format(MILLION_GREY, 5));
});


tests/TestCase.php
<?php

namespace Ahoicloud\Co2\Tests;

use Ahoicloud\Co2\Co2ServiceProvider;
use Illuminate\Database\Eloquent\Factories\Factory;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Ahoicloud\\Co2\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            Co2ServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');

        /*
        $migration = include __DIR__.'/../database/migrations/create_co2_table.php.stub';
        $migration->up();
        */
    }
}


src/Co2.php
<?php

namespace Ahoicloud\Co2;

use Exception;

class Co2
{
    const GIGABYTE = 1000 * 1000 * 1000;

    /**
     * Accept a figure for bytes transferred and return an object representing
     * the share of the total energy use of the entire system, broken down
     * by each corresponding system component
     * Gets emission by Sustainable Web Design
     *   https://sustainablewebdesign.org/calculating-digital-emissions/.
     *
     * @param    $bytes
     * @return array containing the energy in kilowatt hours, keyed by system component
     */
    public function energyPerByteByComponent($bytes)
    {
        $transferedBytesToGb = $bytes / self::GIGABYTE;
        $energyUsage = $transferedBytesToGb * config('co2.KWH_PER_GB');

        // return the total energy, with breakdown by component
        return [
            'consumerDeviceEnergy' => $energyUsage * config('co2.END_USER_DEVICE_ENERGY'),
            'networkEnergy' => $energyUsage * config('co2.NETWORK_ENERGY'),
            'productionEnergy' => $energyUsage * config('co2.PRODUCTION_ENERGY'),
            'dataCenterEnergy' => $energyUsage * config('co2.DATACENTER_ENERGY'),
        ];
    }

  /**
   * Accept a figure for bytes transferred and return the number of kilowatt hours used
   * by the total system for this data transfer
   *
   * @param  float  $bytes
   * @return float the number of kilowatt hours used
   */
  public function energyPerByte(float $bytes)
  {
      $energyByComponent = $this->energyPerByteByComponent($bytes);

      // pull out our values…
      $energyValues = collect($energyByComponent);

      // so we can return their sum
      return $energyValues->sum();
  }

  /**
   * Accept an object keys by the different system components, and
   * return an object with the co2 figures key by the each component
   *
   * @param  array  $energyByComponent - energy grouped by the four system components
   * @param  bool  $carbonIntensity[] - carbon intensity to apply to the datacentre values
   * @return array the total number in grams of CO2 equivalent emissions
   */
  public function co2byComponent($energyByComponent, $carbonIntensity = false)
  {
      $deviceCarbonIntensity = config('co2.GLOBAL_GRID_INTENSITY');
      $networkCarbonIntensity = config('co2.GLOBAL_GRID_INTENSITY');
      $dataCenterCarbonIntensity = config('co2.GLOBAL_GRID_INTENSITY');

      $globalEmissions = config('co2.GLOBAL_GRID_INTENSITY');

      // If the user passes in a TRUE value (green web host), then use the renewables intensity value
      if ($carbonIntensity === true) {
          $dataCenterCarbonIntensity = config('co2.RENEWABLES_GRID_INTENSITY');
      } else {
          $dataCenterCarbonIntensity = config('co2.GLOBAL_GRID_INTENSITY');
      }

      $returnCO2ByComponent = [];
      foreach ($energyByComponent as $key => $value) {
          if (strpos($key, 'dataCenterEnergy') === 0) {
              $returnCO2ByComponent[str_replace('Energy', 'CO2', $key)] = $value * $dataCenterCarbonIntensity;
          } elseif (strpos($key, 'consumerDeviceEnergy') === 0) {
              $returnCO2ByComponent[str_replace('Energy', 'CO2', $key)] = $value * $deviceCarbonIntensity;
          } elseif (strpos($key, 'networkEnergy') === 0) {
              $returnCO2ByComponent[str_replace('Energy', 'CO2', $key)] = $value * $networkCarbonIntensity;
          } else {
              $returnCO2ByComponent[str_replace('Energy', 'CO2', $key)] = $value * $globalEmissions;
          }
      }

      return $returnCO2ByComponent;
  }

    /**
     * Accept a figure for bytes transferred and return a single figure for CO2
     * emissions. Where information exists about the origin data is being
     * fetched from, a different carbon intensity figure
     * is applied for the datacentre share of the carbon intensity.
     *
     * @param    $bytes - the data transferred in bytes
     * @param  bool  $carbonIntensity the carbon intensity for datacentre (average figures, not marginal ones)
     * @param  bool  $segmentResults the result should return as segmentet
     * @return float|null|array the total number in grams of CO2 equivalent emissions
     */
    public function perByte($bytes, $carbonIntensity = false, $segmentResults = false, $options = [])
    {
        $energyBycomponent = $this->energyPerByteByComponent($bytes);

        // otherwise when faced with non numeric values throw an error
        if (! is_bool($carbonIntensity)) {
            throw new Exception(
                'perByte expects a boolean for the carbon intensity value. Received: '.$carbonIntensity
            );
        }

        $co2ValuesbyComponent = $this->co2byComponent(
            $energyBycomponent,
            $carbonIntensity
        );

        // pull out our values…
        $co2Values = collect($co2ValuesbyComponent);
        $co2ValuesSum = $co2Values->sum();

        if ($segmentResults) {
            return collect($co2Values)->merge(['total' => $co2ValuesSum])->all();
        }

        return $co2ValuesSum;
    }
}


config/co2.php
<?php

//Constants taken from https://github.com/thegreenwebfoundation/co2.js/blob/main/src/constants/index.js
// config for Ahoicloud/Co2
return [
    // SUSTAINABLE WEB DESIGN CONSTANTS
    // this refers to the estimated total energy use for the internet around 2000 TWh,
    // divided by the total transfer it enables around 2500 exabytes
    'KWH_PER_GB' => 0.81,
    // these constants outline how the energy is attributed to
    // different parts of the system in the SWD model
    'END_USER_DEVICE_ENERGY' => 0.52,
    'NETWORK_ENERGY' => 0.14,
    'DATACENTER_ENERGY' => 0.15,
    'PRODUCTION_ENERGY' => 0.19,
    // These carbon intensity figures https://ember-climate.org/data/data-explorer
    // - Global carbon intensity for 2021
    'GLOBAL_GRID_INTENSITY' => 442,
    'RENEWABLES_GRID_INTENSITY' => 50,

    // Taken from: https://gitlab.com/wholegrain/carbon-api-2-0/-/blob/master/includes/carbonapi.php

    'FIRST_TIME_VIEWING_PERCENTAGE' => 0.75,
    'RETURNING_VISITOR_PERCENTAGE' => 0.25,
    'PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD' => 0.02,

];
